src/gpt.py

```diff
@@ -0,0 +1,312 @@
+        # No idea if these are sorted by time
+        next_event = message.guild.scheduled_events[0]
+        if ":" in next_event.name:
+            event_num, event_topic = next_event.name.split(":", maxsplit=1)
+            result.append(f"Next bust event: {event_num.strip()}")
+            result.append(f"Next bust topic: {event_topic.strip()}")
+        else:
+            result.append(f"Next bust event: {next_event.name}")
+
+        result.append(f"Next bust time: {next_event.start_time.strftime('%b %d')}")
+        result.append(f"Today's date: {datetime.date.today().strftime('%b %d')}")
+
+    user = get_name(message.author)
+
+    # Role-based info
+    if isinstance(message.author, Member):
+        roles = {role.name for role in message.author.roles}
+
+        # Provide champion info
+        champ = [
+            ("Defending Champion", "first"),
+            ("Runner-up", "second"),
+            ("Bronzer", "third"),
+        ]
+        for role, place in champ:
+            if role in roles:
+                result.append(f"{user}'s place last bust: {place}")
+                break
+
+    result.append(f"Talking to: {user}")
+    return result
+
+
+# Count tokens in a string
+def token_count(data: str) -> int:
+    return len(encoding.encode(data))
+
+
+# Replace Discord-style mentions with names
+def strip_mentions(message: Message) -> str:
+    content = message.content
+    for user in message.mentions:
+        content = content.replace(user.mention, get_name(user))
+    for channel in message.channel_mentions:
+        content = content.replace(channel.mention, channel.name)
+    for role in message.role_mentions:
+        content = content.replace(role.mention, role.name)
+    return content
+
+
+# Fetch messages from history up to a certain token allowance
+async def fetch_history(
+    token_limit: int, speaking_turn_limit: int, message: Message
+) -> List[Tuple[str, bool]]:
+    total_tokens = 0
+    history = []
+    idx = 0
+    one_hour = datetime.timedelta(hours=1)
+    seen_message = False
+    speaking_turn_count = 0
+    last_author = None
+    async for msg in message.channel.history():
+        # Skip messages newer than what we're replying to
+        seen_message = seen_message or (msg == message)
+        if not seen_message:
+            continue
+
+        # Don't include messages which are both more than 3 back and an hour old
+        idx += 1
+        timezone = message.created_at.tzinfo
+        now = datetime.datetime.now(timezone)
+        if idx > 3 and (now - message.created_at) > one_hour:
+            break
+
+        # Break if maximum number of speaking turns has been reached
+        if msg.author != last_author:
+            speaking_turn_count += 1
+            if speaking_turn_count > speaking_turn_limit:
+                break
+            last_author = msg.author
+
+        # Break if message is disallowed
+        if disallowed_message(msg):
+            break
```

Comment:
I was thinking that 
1. if these terms are being thrown around Busty might be in "LLM Mode" and need a full reset
2. Context might not "make sense" with stuff missing, though i think chatgpt is not really trained to call out no-sense convos so this might not be an issue

I'd be ok with just skipping

Reply:
I suggest trying it briefly and seeing if it helps. Most of this is experimentation anyhow :art:

Reply:
I suggest trying it briefly and seeing if it helps. Most of this is experimentation anyways :art:

Reply:
I tried a number of things and while I got her to admit to being an LLM a few times, I definitely used some knowledge of specific trigger words to get around and stuff. I think just skipping bad messages instead of breaking on them is good enough.

![image](https://user-images.githubusercontent.com/6956898/228428636-a0f7c158-beab-4261-a952-535d7b799406.png)